hypothesis-python/src/hypothesis/internal/conjecture/engine.py

```diff
@@ -310,15 +310,13 @@ def __init__(
 
         self.interesting_examples: dict[InterestingOrigin, ConjectureResult] = {}
         # We use call_count because there may be few possible valid_examples.
-        self.first_bug_found_at: Optional[int] = None
-        self.last_bug_found_at: Optional[int] = None
+        self.first_bug_found_at: int | None = None
+        self.last_bug_found_at: int | None = None
 
         self.shrunk_examples: set[InterestingOrigin] = set()
-        self.health_check_state: Optional[HealthCheckState] = None
+        self.health_check_state: HealthCheckState | None = None
         self.tree: DataTree = DataTree()
-        self.provider: Union[type, PrimitiveProvider] = _get_provider(
-            self.settings.backend
-        )
+        self.provider: type | PrimitiveProvider = _get_provider(self.settings.backend)
```

Comment:
```suggestion
        self.provider: type[PrimitiveProvider] | PrimitiveProvider = _get_provider(self.settings.backend)
```
~~or perhaps `Callable[[], PrimitiveProvider]`~~ no, we use `type[PrimitiveProvider]` elsewhere